File: nwerc2024/evolvingetymology/submissions/time_limit_exceeded/ragnar_linear.py
#!/usr/bin/env python3

n, k = map(int, input().split())
name = input()
stride = 1

for _ in range(k):
    stride *= 2
    stride -= n * (stride >= n)
pos = 0
for i in range(n):
    print(name[pos], end="")
    pos += stride
    pos -= n * (pos >= n)
print()

File: nwerc2024/evolvingetymology/submissions/time_limit_exceeded/ragnar_linear.cpp
#include <iostream>
using namespace std;
int main() {
	int n, k;
	cin >> n >> k;
	string name;
	cin >> name;
	int stride = 1;
	for(int i = 0; i < k; ++i) {
		stride *= 2;
		stride -= n * (stride >= n);
	}
	int pos = 0;
	for(int i = 0; i < n; ++i) {
		cout << name[pos];
		pos += stride;
		pos -= n * (pos >= n);
	}
	cout << endl;
}
